### ISSUE TITLE
PaperTrail::Serializers::JSON not compatible with Postgres arrays in Rails >= 5.0.2

### ISSUE THREAD

**Issue by hubertpompecki** [airblade/paper_trail#1015]
Hi everyone,

After updating Rails version from 5.0.1 to 5.0.6 I started experiencing issues when reifying attributes persisted in PostgreSQL arrays. It appears that data gets corrupted in the serialization process when building a review body and comes back as `nil` when restored.

I believe this is due to the change of the return value of `ActiveRecord::ConnectionAdapters::PostgreSQL::OID::Array#serialize` introduced in Rails 5.0.2. In <= 5.0.1:
```
ActiveRecord::ConnectionAdapters::PostgreSQL::OID::Array.serialize([1,2,3]) => "{1,2,3}"
```
In >= 5.0.2:
```
ActiveRecord::ConnectionAdapters::PostgreSQL::OID::Array.serialize([1,2,3]) => #<struct ActiveRecord::ConnectionAdapters::PostgreSQL::OID::Array::Data encoder=#<PG::TextEncoder::Array:0x007fb36f2b9c18 "integer[]"  elements_type=nil needs quotation>, values=[1, 2, 3]>
```

The return value is now a struct which does not get correctly serialized/deserialized by `ActiveSupport::JSON` to which `PaperTrail::Serializers::JSON` delegates.

If this is indeed a bug and not me missing something please let me know if you have any preferred solutions - I am happy to work on a fix for this. 

## Reproduction
```
require "bundler/inline"

gemfile(true) do
  ruby "2.4.2"
  source "https://rubygems.org"
  gem "activerecord", "5.0.2"
  gem "minitest", "5.10.3"
  gem "paper_trail", "8.0.0", require: false
  gem "pg", "~> 0.21"
end

require "active_record"
require "minitest/autorun"
require "logger"

ActiveRecord::Base.establish_connection(adapter: "postgresql", database: "paper_trail_bug")
ActiveRecord::Base.logger = nil
ActiveRecord::Schema.define do
  create_table :users, force: true do |t|
    t.text :first_name, null: false
    t.integer :post_ids, default: [], array: true
    t.timestamps null: false
  end

  create_table :versions, force: true do |t|
    t.string :item_type, null: false
    t.integer :item_id, null: false
    t.string :event, null: false
    t.string :whodunnit
    t.text :object, limit: 1_073_741_823
    t.text :object_changes, limit: 1_073_741_823
    t.integer :transaction_id
    t.datetime :created_at
  end
  add_index :versions, %i[item_type item_id]
  add_index :versions, [:transaction_id]

  create_table :version_associations, force: true do |t|
    t.integer  :version_id
    t.string   :foreign_key_name, null: false
    t.integer  :foreign_key_id
  end
  add_index :version_associations, [:version_id]
  add_index :version_associations, %i[foreign_key_name foreign_key_id],
    name: "index_version_associations_on_foreign_key"
end
ActiveRecord::Base.logger = Logger.new(STDOUT)
require "paper_trail/config"

require "paper_trail"
PaperTrail::Config.instance.track_associations = true
PaperTrail.serializer = PaperTrail::Serializers::JSON

class User < ActiveRecord::Base
  has_paper_trail
end

class BugTest < ActiveSupport::TestCase
  def test_array_serializing
    user = User.create(first_name: "Josh", post_ids: [1,2,3])
    user.update_attributes(post_ids: [4,5,6])
    previous_user = user.paper_trail.previous_version
    ActiveRecord::Base.logger.debug("Restored post_ids: #{previous_user.post_ids.inspect}")
    assert previous_user.post_ids == [1,2,3]
  end
end
```

## Output
```
Fetching gem metadata from https://rubygems.org/...........
Fetching gem metadata from https://rubygems.org/.
Resolving dependencies...
Using concurrent-ruby 1.0.5
Using i18n 0.9.1
Using minitest 5.10.3
Using thread_safe 0.3.6
Using tzinfo 1.2.4
Fetching activesupport 5.0.2
Installing activesupport 5.0.2
Fetching activemodel 5.0.2
Installing activemodel 5.0.2
Fetching arel 7.1.4
Installing arel 7.1.4
Fetching activerecord 5.0.2
Installing activerecord 5.0.2
Using bundler 1.16.0
Using request_store 1.3.2
Using paper_trail 8.0.0
Using pg 0.21.0
-- create_table(:users, {:force=>true})
   -> 0.0396s
-- create_table(:versions, {:force=>true})
   -> 0.0056s
-- add_index(:versions, [:item_type, :item_id])
   -> 0.0035s
-- add_index(:versions, [:transaction_id])
   -> 0.0024s
-- create_table(:version_associations, {:force=>true})
   -> 0.0063s
-- add_index(:version_associations, [:version_id])
   -> 0.0027s
-- add_index(:version_associations, [:foreign_key_name, :foreign_key_id], {:name=>"index_version_associations_on_foreign_key"})
   -> 0.0021s
Run options: --seed 56854

# Running:

D, [2017-12-01T12:07:53.244737 #70975] DEBUG -- :   [1m[35m (0.1ms)[0m  [1m[35mBEGIN[0m
D, [2017-12-01T12:07:53.246870 #70975] DEBUG -- :   [1m[35mSQL (0.5ms)[0m  [1m[32mINSERT INTO "users" ("first_name", "post_ids", "created_at", "updated_at") VALUES ($1, $2, $3, $4) RETURNING "id"[0m  [["first_name", "Josh"], ["post_ids", #<struct ActiveRecord::ConnectionAdapters::PostgreSQL::OID::Array::Data encoder=#<PG::TextEncoder::Array:0x00007ffc719707f0 "integer[]"  elements_type=nil needs quotation>, values=[1, 2, 3]>], ["created_at", 2017-12-01 12:07:53 UTC], ["updated_at", 2017-12-01 12:07:53 UTC]]
D, [2017-12-01T12:07:53.267491 #70975] DEBUG -- :   [1m[35mSQL (1.0ms)[0m  [1m[32mINSERT INTO "versions" ("item_type", "item_id", "event", "object_changes", "created_at") VALUES ($1, $2, $3, $4, $5) RETURNING "id"[0m  [["item_type", "User"], ["item_id", 1], ["event", "create"], ["object_changes", "{\"id\":[null,1],\"first_name\":[null,\"Josh\"],\"post_ids\":[{\"encoder\":null,\"values\":null},{\"encoder\":1,\"values\":2}],\"created_at\":[null,\"2017-12-01T12:07:53.245Z\"],\"updated_at\":[null,\"2017-12-01T12:07:53.245Z\"]}"], ["created_at", 2017-12-01 12:07:53 UTC]]
D, [2017-12-01T12:07:53.269511 #70975] DEBUG -- :   [1m[35mSQL (0.4ms)[0m  [1m[33mUPDATE "versions" SET "transaction_id" = $1 WHERE "versions"."id" = $2[0m  [["transaction_id", 1], ["id", 1]]
D, [2017-12-01T12:07:53.270294 #70975] DEBUG -- :   [1m[35m (0.4ms)[0m  [1m[35mCOMMIT[0m
D, [2017-12-01T12:07:53.270614 #70975] DEBUG -- :   [1m[35m (0.1ms)[0m  [1m[35mBEGIN[0m
D, [2017-12-01T12:07:53.272765 #70975] DEBUG -- :   [1m[35mSQL (0.5ms)[0m  [1m[33mUPDATE "users" SET "post_ids" = $1, "updated_at" = $2 WHERE "users"."id" = $3[0m  [["post_ids", #<struct ActiveRecord::ConnectionAdapters::PostgreSQL::OID::Array::Data encoder=#<PG::TextEncoder::Array:0x00007ffc719707f0 "integer[]"  elements_type=nil needs quotation>, values=[4, 5, 6]>], ["updated_at", 2017-12-01 12:07:53 UTC], ["id", 1]]
D, [2017-12-01T12:07:53.276046 #70975] DEBUG -- :   [1m[35mSQL (0.3ms)[0m  [1m[32mINSERT INTO "versions" ("item_type", "item_id", "event", "object", "object_changes", "created_at") VALUES ($1, $2, $3, $4, $5, $6) RETURNING "id"[0m  [["item_type", "User"], ["item_id", 1], ["event", "update"], ["object", "{\"id\":1,\"first_name\":\"Josh\",\"post_ids\":{\"encoder\":1,\"values\":2},\"created_at\":\"2017-12-01T12:07:53.245Z\",\"updated_at\":\"2017-12-01T12:07:53.245Z\"}"], ["object_changes", "{\"post_ids\":[{\"encoder\":1,\"values\":2},{\"encoder\":4,\"values\":5}],\"updated_at\":[\"2017-12-01T12:07:53.245Z\",\"2017-12-01T12:07:53.271Z\"]}"], ["created_at", 2017-12-01 12:07:53 UTC]]
D, [2017-12-01T12:07:53.277290 #70975] DEBUG -- :   [1m[35mSQL (0.4ms)[0m  [1m[33mUPDATE "versions" SET "transaction_id" = $1 WHERE "versions"."id" = $2[0m  [["transaction_id", 2], ["id", 2]]
D, [2017-12-01T12:07:53.277961 #70975] DEBUG -- :   [1m[35m (0.3ms)[0m  [1m[35mCOMMIT[0m
D, [2017-12-01T12:07:53.289300 #70975] DEBUG -- :   [1m[36mPaperTrail::Version Load (0.3ms)[0m  [1m[34mSELECT  "versions".* FROM "versions" WHERE "versions"."item_id" = $1 AND "versions"."item_type" = $2 ORDER BY "versions"."created_at" DESC, "versions"."id" DESC LIMIT $3[0m  [["item_id", 1], ["item_type", "User"], ["LIMIT", 1]]
D, [2017-12-01T12:07:53.294400 #70975] DEBUG -- :   [1m[36mUser Load (0.2ms)[0m  [1m[34mSELECT  "users".* FROM "users" WHERE "users"."id" = $1 LIMIT $2[0m  [["id", 1], ["LIMIT", 1]]
D, [2017-12-01T12:07:53.296684 #70975] DEBUG -- : Restored post_ids: nil
F

Finished in 0.063553s, 15.7349 runs/s, 15.7349 assertions/s.

  1) Failure:
BugTest#test_array_serializing [my_bug_report.rb:72]:
Expected false to be truthy.

1 runs, 1 assertions, 1 failures, 0 errors, 0 skips
```

**Comment by jaredbeck** [comment]
Thanks for the excellent bug report, Hubert. I can reproduce the issue using your script.

This does not appear to be an issue for columns with datatype json or jsonb. As a sidenote, we recommend that postgres users convert their `object` and `object_changes` columns to json(b). https://github.com/airblade/paper_trail/#postgresql-json-column-type-support

I noticed that this is less of an issue for YAML, because it stores something like:

```yaml
---
id: 1
first_name: Josh
post_ids: !ruby/struct:ActiveRecord::ConnectionAdapters::PostgreSQL::OID::Array::Data
  encoder: !ruby/object:PG::TextEncoder::Array
    name: integer[]
    elements_type: 
  values:
  - 1
  - 2
  - 3
```

That works, but it's not great. I'd rather store a plain array. We should fix this for both serializers, if possible.

> I am happy to work on a fix for this.

Thanks. You could start with a test like this, perhaps?

```
describe ObjectAttribute do
  describe '#serialize' do
    it 'serializes array into plain array' do
      atr = { 'post_ids' => [1, 2, 3] }
      PaperTrail::AttributeSerializers::ObjectAttribute.new(User).serialize(atr)
      assert_equal [1, 2, 3], atr['post_ids']
    end
```

**Comment by jaredbeck** [comment]
Closed by #1018
